sdk/cosmos/azure_data_cosmos/src/retry_policies/client_retry_policy.rs

```diff
@@ -0,0 +1,1130 @@
+
+            RetryResult::Retry {
+                after: Duration::ZERO,
+            }
+        }
+    }
+
+    /// Determines if a request should be retried when an endpoint fails.
+    ///
+    /// # Summary
+    /// Handles endpoint failures by marking failed endpoints as unavailable and attempting retry
+    /// on alternative endpoints. Refreshes the location cache to get updated endpoint information
+    /// and configures retry delays based on request type (write requests get longer delays).
+    /// Respects maximum retry limits and endpoint discovery settings. Can mark endpoints as
+    /// unavailable for reads, writes, or both depending on the failure scenario.
+    ///
+    /// # Arguments
+    /// * `is_read_request` - Whether this is a read operation
+    /// * `mark_both_read_and_write_as_unavailable` - Whether to mark the endpoint unavailable for both operations
+    /// * `force_refresh` - Whether to force refresh of the location cache
+    /// * `retry_on_preferred_locations` - Whether to retry on preferred locations first
+    /// * `overwrite_endpoint_discovery` - Whether to bypass endpoint discovery checks
+    ///
+    /// # Returns
+    /// A `RetryResult`:
+    /// - `Retry { after: Duration }` with appropriate delay if retry is allowed
+    /// - `DoNotRetry` if max retry count exceeded or endpoint discovery disabled
+    async fn should_retry_on_endpoint_failure_async(
+        &mut self,
+        is_read_request: bool,
+        mark_both_read_and_write_as_unavailable: bool,
+        force_refresh: bool,
+        retry_on_preferred_locations: bool,
+        overwrite_endpoint_discovery: bool,
+    ) -> RetryResult {
+        if self.failover_retry_count > MAX_RETRY_COUNT_ON_ENDPOINT_FAILURE
+            || (!self.enable_endpoint_discovery && !overwrite_endpoint_discovery)
+        {
+            return RetryResult::DoNotRetry;
+        }
+
+        self.failover_retry_count += 1;
+
+        if let Some(ref endpoint) = self.location_endpoint {
+            if !overwrite_endpoint_discovery {
+                if is_read_request || mark_both_read_and_write_as_unavailable {
+                    self.global_endpoint_manager
+                        .mark_endpoint_unavailable_for_read(endpoint);
+                }
+                if !is_read_request || mark_both_read_and_write_as_unavailable {
+                    self.global_endpoint_manager
+                        .mark_endpoint_unavailable_for_write(endpoint);
+                }
+            }
+        }
+
+        let retry_delay = if !is_read_request {
+            if self.failover_retry_count > 1 {
+                Duration::milliseconds(RETRY_INTERVAL_MS)
+            } else {
+                Duration::ZERO
+            }
+        } else {
+            Duration::milliseconds(RETRY_INTERVAL_MS)
+        };
+
+        let _refresh_cache_result = self
+            .global_endpoint_manager
+            .refresh_location_async(force_refresh)
+            .await;
```

Comment:
Another lost error here.

Reply:
Yes, typically the outcome should be ignored, because we want the request to still go through, when the cache fails to refresh due to some intermittent connectivity issue. Added a detailed code comment to explain the scenario.